Implement CSP check for Trusted Types

TimvdLippe opened the pull request:
The algorithm [1] is implemented in the content-security-policy
package.

Requires https://github.com/rust-ammonia/rust-content-security-policy/pull/56
This is part of #36258

[1]: https://w3c.github.io/trusted-types/dist/spec/#abstract-opdef-should-trusted-type-policy-creation-be-blocked-by-content-security-policy

TimvdLippe commented:
This is work-in-progress, which requires an update to an upstream package `content-security-policy`. Unfortunately local testing shows that this isn't working as expected yet. It does correctly throw a type error (`Not allowed by CSP`), but it does not dispatch the `SecurityPolicyViolationEvent`. Therefore, the tests are failing as they assert that the event is thrown, rather than that the `TypeError` is thrown.

I am not sure how to proceed, as I didn't see pre-existing usages of the CSPViolationsReporter. I did see some examples where there is relevant code, but they are explicitly commented with `// TODO`:

https://github.com/servo/servo/blob/3f24b44e1581af01a3fb51c031663d1dd99a790c/components/script/dom/document.rs#L3996-L4002

I also am not sure how to document the algorithm correctly. As you can see, step 2.11 is executed in Servo, whereas all the other steps are executed in the `content-security-policy` package. This makes documenting a bit awkward, as it is only partly true that the reference to step 2.11 in `content-security-policy` implements the relevant parts.

Let me know what your thoughts are and if you have any guidance on how to proceed.

sebsebmc commented:
For uses of `CSPViolationReporter` take a look at 
https://github.com/servo/servo/blob/b87bf0b806a0dd21136faf04db47af4cb356922e/components/script/script_runtime.rs#L403
and https://github.com/servo/servo/blob/b87bf0b806a0dd21136faf04db47af4cb356922e/components/script/dom/globalscope.rs#L2681

TimvdLippe commented:
@sebsebmc yes thanks I took those as inspiration and that's why I copied these lines thinking it would fire these events. However, when I run the tests I don't see any events being fired. Did I copy the lines wrong?

jdm commented:
I don't think your issue is with the usage of CSPViolationReporter—I think it's caused by parsing issues in rust-content-security-policy that mean that no violations are returned by that library. In particular, this unit test fails right now:
```
    #[test]
    pub fn basic_policy_is_valid2() {
        let p = Policy::parse("trusted-types 'none'", PolicySource::Meta, PolicyDisposition::Enforce);
        assert!(p.is_valid());
        assert_eq!(p.directive_set[0].value, vec!["none".to_owned()]);
    }
```
The parsed value is `'none'` instead of `none`.

jdm commented:
Looking at the code some more, existing code specifically checks for single values in surrounding quotes, so I guess "parsing bug" isn't exactly right. The PR won't work as written, however.

jdm commented:
Looking at the test code, the timeouts occur because we never fire a securitypolicyviolation event with the connect-src effective direction, which apparently is used as a signal that there will be no further tests: https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/tests/wpt/tests/trusted-types/support/csp-violations.js#L24-L27

jdm commented:
Oh, here it is: https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/tests/wpt/tests/trusted-types/support/csp-violations.js#L40-L49

jdm commented:
Servo doesn't implement any CSP checks for connect-src yet, so that's going to block this work, unfortunately.

TimvdLippe commented:
Thanks for the investigation, that helps me a lot to understand what's going on. The reason I thought it was related to the event, is that in the Console when debugging with Firefox DevTools it correctly showed the TypeError that it wasn't allowed by CSP. That's why I concluded that my CSP parsing logic was working, but it was related to the event.

I hadn't realized that `connect-src` isn't implemented yet. I will take a stab at that and see if I can make that work. In the other Rust package I saw pre-existing support for it, so maybe it is more straightforward than I would fear atm. Will report back 👍

jdm commented:
To be clear, I'm open to merging PRs that implement just enough of the network CSP checks to allow this work to continue. I don't want to stop you from contributing to the features that you're actually interested in working on :)

TimvdLippe commented:
For sure! I found out that in the following code we have the same TODO:

https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/net/fetch/methods.rs#L193-L198

I am currently trying to figure out how many changes are required to make that work and if it makes the Trusted Types tests pass. Will let you know, thanks for all the guidance!

jdm commented:
Be aware that that code probably won't be invoked for websockets until #35028 is fixed. Instead we go through https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/net/websocket_loader.rs#L347

jdm commented:
Adding a new method to https://doc.servo.org/net_traits/trait.FetchResponseListener.html#tymethod.process_response for handling CSP violations might be the way to go. They could be stored in a new variant of the NetworkError enum.

TimvdLippe commented:
Note-taking for myself:

https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/shared/net/request.rs#L125-L130
We are missing the client object here, as defined in https://fetch.spec.whatwg.org/#concept-request-window I assume that we need a third enum option to store the global here.

https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/script/dom/request.rs#L217
We need to set the value here
https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/script/dom/request.rs#L163
And this method has the `GlobalScope` as required

However, we can't add `GlobalScope` as an import dependency to `components/shared/net/request.rs` since it lives in the `script` package and the `request.rs` in `net_traits`. Since `script` already depends on `net_traits`, we can't add a circular dependency here. Also, I assume that we want to keep `net_traits` lean (since it is a `_traits`) package, thus we shouldn't depend on a heavy object there.

Two options that come to mind:
1. Define a custom trait (here my Rust knowledge is lacking, but I think this is an "interface" as I know it) that both packages can depend on
2. We let `net_traits` depend on `script_bindings`, since we can use the generated `GlobalScope` based on the `.webidl` file

Neither options are great, since it would mean that `net_traits` would gain a large dependency tree.

Alternatively, we can add a callback to to `components/shared/net/request.rs#Window` which we then invoke and we implement the reporting in `components/script/dom/request.rs` instead.

I will start with that last one and see how far I can get.

TimvdLippe commented:
All right, that doesn't work, since the enum needs to be cloneable etc..

```rs
type GlobalCallback = Box<dyn Fn(Vec<Violation>)>;

/// [Window](https://fetch.spec.whatwg.org/#concept-request-window)
#[derive(Clone, Copy, MallocSizeOf, PartialEq)]
pub enum Window {
    NoWindow,
    Client,
    Global {
        global_callback: GlobalCallback
    }
}
```

And since the callback would need access to `&global`, that is not allowed:

```rs
let global_callback = |violations: Vec<Violation>| {
    for violation in violations {
        let task = CSPViolationReporter::new(
            &global,
            None,
            false,
            RuntimeCode::JS,
            "".to_string(),
            0,
            0,
        );
        &global
            .task_manager()
            .dom_manipulation_task_source()
            .queue(task);
    }
};
// Step 8
let mut window = Window::Global {
    global_callback,
};
```

Here my lack of Rust experience is shining, I need to learn more about these traits and what idiomatic Rust is to achieve this.

> Adding a new method to https://doc.servo.org/net_traits/trait.FetchResponseListener.html#tymethod.process_response for handling CSP violations might be the way to go. They could be stored in a new variant of the NetworkError enum.

If I understand correctly, you are referring to this part of code: https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/net/fetch/methods.rs#L281-L286

Here, if I understand you correctly, you would add a new enum value that would contain the `violations` as captured. However, that would mean that we only report the violations when it would be blocked. We would also need to report the violations when we have `CheckResult::Allowed` but `PolicyDisposition::Report`. Therefore, I don't think we can capture this in a `NetworkError` 😢

simonwuelker commented:
> Alternatively, we can add a callback to to components/shared/net/request.rs#Window which we then invoke and we implement the reporting in components/script/dom/request.rs instead.

You don't need a callback, at least not `type GlobalCallback = Box<dyn Fn(Vec<Violation>)>;`.

What you are describing is essentially the [`FetchResponseListener`](https://doc.servo.org/net_traits/trait.FetchResponseListener.html) trait. When some code wants to do a fetch operation it will pass a type that implements [`FetchResponseListener`](https://doc.servo.org/net_traits/trait.FetchResponseListener.html) to the fetch code, which will then invoke the appropriate methods as the fetch is executed. (`process_request_eof` when the request has been fully sent, `process_response_chunk` whenever some new data from the response body is available etc).

For example, this is how `<img>` tags load in servo: https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/script/dom/htmlimageelement.rs#L215

On the fetch side of things you have a `Target`, which is *something* that implements [`FetchTaskTarget`](https://doc.servo.org/net_traits/trait.FetchTaskTarget.html)

https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/net/fetch/methods.rs#L52

[`FetchTaskTarget`](https://doc.servo.org/net_traits/trait.FetchTaskTarget.html) is more or less the same thing as [`FetchResponseListener`](https://doc.servo.org/net_traits/trait.FetchResponseListener.html), except that it takes care of ipc.

https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/shared/net/lib.rs#L215-L238

https://github.com/servo/servo/blob/9d074161638829deb0f0b1adacb8ff270abd1978/components/shared/net/lib.rs#L287-L317

So as jdm mentioned you will probably want to add a method like `process_csp_violation(s)` to both [`FetchResponseListener`](https://doc.servo.org/net_traits/trait.FetchResponseListener.html) and [`FetchTaskTarget`](https://doc.servo.org/net_traits/trait.FetchTaskTarget.html).

> Therefore, I don't think we can capture this in a `NetworkError`

Right, but I don't think you need to. It should be possible to report the violations immediately by calling the appropriate method on `target` and not touch the rest of the method at all.

TimvdLippe commented:
Pushed some work-in-progress stuff. I tried to run it with

```
./mach test-wpt tests/wpt/tests/trusted-types/should-trusted-type-policy-creation-be-blocked-by-csp-001.html
```

This is a test that uses an `iframe` to test it. Unfortunately that one doesn't have a `FetchResponseListener` itself. I thought `ScriptThread` would be the one, but probably not. Need to go now, so will continue investigating later.

jdm commented:
Mm, yeah, all navigations have a listener in the Constellation instead (components/constellation) since they may end up being route to a different script thread than the originator if they are cross-origin.

jdm commented:
Nevermind, I am remembering an old state of the code. https://github.com/servo/servo/issues/23037 shows that we never handled that case correctly, and https://github.com/servo/servo/pull/34919 moved all navigation out of the constellation into the script thread. In any case, all the relevant logic is now in components/script/navigation.rs.

TimvdLippe commented:
Implemented all the relevant implementations of the listener. Unfortunately the WPT test still fails, but haven't had the time to investigate further as to why. Also, I want to write some dummy tests myself to see if at least the Trusted Types CSP part is working as I expect.

PTAL and let me know if I implemented it correctly as you suggested.

TimvdLippe commented:
I also started a try run to see if other WPT tests related to CSP are passing now: https://github.com/TimvdLippe/servo/actions/runs/14302153705

TimvdLippe commented:
Correct try run by specifying the dependency via `git` instead of `path` so that it can run on CI: https://github.com/TimvdLippe/servo/actions/runs/14302256149

jdm commented:
Very nice! The missing piece now is a custom integration with the websockets code, which is what the trusted types code relies upon.

TimvdLippe commented:
Relevant WPT test expectations have been updated, running final verification check in https://github.com/TimvdLippe/servo/actions/runs/14327159924

The lint CI check will fail, as I still rely on my git branch for https://github.com/rust-ammonia/rust-content-security-policy/pull/56 When that's released we can update the Cargo dependency.

Please review the other bits of the code and let me know if I missed anything. FYI I expect many more Trusted Types CSP tests to pass once we fix the Websocket implementation. However, I did spot an issue with the `effective_directive` set to `script-src` for all violations. That needs to be fixed as well, but the tests appear to pass because it is included in the `cspDirectivses`:

https://github.com/servo/servo/blob/40655cc06cbeb9affd8f4eeae9f45cd57db319f1/tests/wpt/tests/trusted-types/support/csp-violations.js#L6-L7
https://github.com/servo/servo/blob/40655cc06cbeb9affd8f4eeae9f45cd57db319f1/tests/wpt/tests/trusted-types/support/csp-violations.js#L22

jdm commented:
Looks like the policy container will need to be set in https://github.com/servo/servo/blob/main/components/script/dom/websocket.rs#L261 as well.

TimvdLippe commented:
@jdm Sure thing, here it is: #36409 Unfortunately it also requires us updating the csp crate. And the tests don't pass and I don't see why they wouldn't. I think we are still missing parts of the puzzle here.

jdm committed:
The algorithm [1] is implemented in the content-security-policy package.

This is part of #36258

[1]: https://w3c.github.io/trusted-types/dist/spec/#abstract-opdef-should-trusted-type-policy-creation-be-blocked-by-content-security-policy

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
This PR requires #36409 merging first.

TimvdLippe commented on tests/wpt/meta/trusted-types/trusted-types-navigation.html.ini:
```diff
@@ -1,5 +1,5 @@
 [trusted-types-navigation.html]
-  expected: TIMEOUT
+  expected: CRASH
```
We run into https://github.com/w3c/webappsec-csp/issues/704 and https://github.com/w3c/trusted-types/issues/577 Seems like the spec is unresolved, so I think it is okay to crash for now.

TimvdLippe commented on tests/wpt/meta/trusted-types/trusted-types-reporting-clipping-of-sample.html.ini:
```diff
@@ -1,37 +1,37 @@
 [trusted-types-reporting-clipping-of-sample.html]
-  expected: TIMEOUT
+  expected: CRASH
```
We run into https://github.com/w3c/webappsec-csp/issues/704 and https://github.com/w3c/trusted-types/issues/577 Seems like the spec is unresolved, so I think it is okay to crash for now.

jdm replied:
Let's file a followup issue for this so we track it!

TimvdLippe commented on tests/wpt/meta/trusted-types/trusted-types-reporting-clipping-of-sample.tentative.html.ini:
```diff
@@ -1,25 +1,25 @@
 [trusted-types-reporting-clipping-of-sample.tentative.html]
-  expected: TIMEOUT
+  expected: CRASH
```
We run into https://github.com/w3c/webappsec-csp/issues/704 and https://github.com/w3c/trusted-types/issues/577 Seems like the spec is unresolved, so I think it is okay to crash for now.

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#51961) with upstreamable changes.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51961).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51961).

TimvdLippe commented:
Checking current progress with a try run: https://github.com/TimvdLippe/servo/actions/runs/14431081392

servo-wpt-sync commented:
🤖 This change no longer contains upstreamable changes to WPT; closed existing upstream pull request (web-platform-tests/wpt#51961).

TimvdLippe commented:
With the other CSP PR merged, this is now ready for review 🎉

jdm reviewed (approved):
This change ended up being so small :D
